esphome/components/json/json_util.cpp

```diff
@@ -51,13 +51,13 @@ void parse_json(const std::string &data, const json_parse_t &f) {
 #ifdef USE_ESP8266
   const size_t free_heap = ESP.getMaxFreeBlockSize();  // NOLINT(readability-static-accessed-through-instance)
 #elif defined(USE_ESP32)
-  const size_t free_heap = heap_caps_get_largest_free_block(MALLOC_CAP_8BIT);
+  const size_t free_heap = heap_caps_get_largest_free_block(MALLOC_CAP_INTERNAL);
 #endif
   bool pass = false;
-  size_t request_size = std::min(free_heap, (size_t)(data.size() * 1.5));
+  size_t request_size = std::min(free_heap - 2048, (size_t)(data.size() * 1.5));
   do {
     DynamicJsonDocument json_document(request_size);
-    if (json_document.capacity() == 0) {
+    if (json_document.memoryPool().buffer() == nullptr) {
```

Comment:
```suggestion
    iif (json_document.capacity() == 0) {
```

We don't need to revert this